Creates test for observationunits call

reynoldtan opened the pull request:
Creates unit test for observationunits call

reynoldtan committed:
Creates test for observationunits call

laceysanderson commented on calls/v1/observationunits/TripalWebServiceBrapiV1Observationunits.inc:
```diff
@@ -146,6 +151,38 @@ class TripalWebServiceBrapiV1Observationunits extends TripalWebServiceDatabaseCa
+    if (isset($call_asset['parameter']['trialDbId'])
+        && !empty($call_asset['parameter']['trialDbId'])) {
+
+      $filter[] = "CONCAT(year.value, '-', loc.value, '-', rep.value, '-', germ.name, '-', unit.name) = :locationDbId";
```
This appears to be a copy/paste from location and doesn't actually filter for trialDbId.

reynoldtan replied:
Good catch!

laceysanderson reviewed (changes requested):
Code Changes:
 - Not all parameter implemented. Missing trialDbId, programDbId, seasonDbId, observationLevel, observationTimeStampRangeStart, observationTimeStampRangeEnd ❌ 
 - are studyLoactionDbId, trailDbId and locationDbId all the same value? ❓ 
Check in the Browser:
 - Structure matches Apiary ✅ 
 - Data correct ✅ 
 - Returns results using test data ✅ 
 - Check parameters:
     - germplasmDbId ✅ 
     - observationVariableDbId ✅ 
     - studyDbId ✅ 
     - locationDbId ✅ 
     - Same missing as mentioned above.
Test Review:
 - Tests all parameters: did not test observationVariableDbId even though it is implemented. ❌ 
 - Includes correct data structure and data types based on Apiary ✅ 
Tests Pass ✅ 

Regarding the unimplemented parameters, it is mentioned in the comment that it was due to an unclear definition. My guess is this is due to the parameter not being in the result. You likely just have to follow it through a relationship. For example, programDbId you will need to filter the results based on studyDbIds which are part of that programDbId.

reynoldtan commented:
Your guess is correct, all calls usually have clearly defined parameter that will always match a field in the response field. In this case I could not establish that link. This is clear though in v2 https://app.swaggerhub.com/apis/PlantBreedingAPI/BrAPI-Phenotyping/2.0#/Observation%20Units/get_observationunits here you can see those parameters now match a response field. We now have 2 different versions of response field - your thoughts? If you ask me, I would follow the v1.3 response field definition and not implement those parameters for transparency reason (parameter can be seen/identified in the response).

reynoldtan commented:
With the exception of the last two parameters about timestamps where I could not find a relevant chado field. Values for these two is always set to n/a.

reynoldtan committed:
Removes an unecessary code block

laceysanderson commented:
I feel those parameters should be implemented in our module for v1.3 calls. 

This is because those parameters are shown available in the v1.3 specification. I hear your argument and agree that it is unfortunate that there are parameters which do not reference a result field but we should follow the specification regardless of our own feelings. Otherwise, we are not implementing BrAPI v1.3 but rather our own interpretation of it.

reynoldtan commented:
No, we are not basing anything on feelings - it is the consistent pattern among all the calls that parameters can be crossed reference against response field that I deemed specs of response field of high priority over parameter. I will implement these parameter but I would need to amend the response field to include these parameters, this way we can run our test on each. What about the timestamps?

laceysanderson commented:
> I will implement these parameter but I would need to amend the response field to include these parameters, this way we can run our test on each.

No, when implementing a specification it should be matched exactly. Do not amend the response field to include these parameters. 

When testing you can use your understanding of the data and/or other calls to test it. For example, programDbId you will need to use the studies call to determine the array of studyDbId for that programDbId and then check that your results only have studyDbId in that array.

laceysanderson commented:
> What about timestamp?

It would be a phenotypeprop which neither KnowPulse or DivSeek Canada Portal has an example of. Again, find a term using EBI and implement it.

For the following:
```
        'n/a' AS plantnumber,
        'n/a' AS plotnumber,
        'n/a' AS positioncoordinatex,
        'n/a' AS positioncoordinatextype,
        'n/a' AS positioncoordinatey,
        'n/a' AS positioncoordinateytype,
```
They should also be implemented as phenotypeprop.

reynoldtan committed:
Revised observationunits call and unit test

reynoldtan commented:
Revised observation call particularly the query, the previous showed incorrect response where each phenotype (observation) is a single unit in the response. This update will now group observations based on study. I have also implemented all parameters. Please note the timestamp is in YYYY-MM-DD HH:MM format so an example of range as parameter for this call might be 

`host/web-services/brapi/v1/observationunits?observationTimeStampRangeStart=2011-06-01 22:06&observationTimeStampRangeEnd=2011-07-31 22:06
`

Other updates:
1. Revised raw csv file for phenotypes.
2. Added a function in unit test that will fetch a field value to allow test of other parameters.
3. Added a parameter data type in core call class.

reynoldtan committed:
Merge branch 'master' into calls-observationunits

reynoldtan committed:
Fixes pdo exception in observationunits call

laceysanderson reviewed (approved):
Changes look good 👍 The previously missing parameters are implemented, have test data and work through the UI. Tests pass ✅  I feel this is ready to merge! 🎉
